File: src/services/api/decisionReview.ts
/**
 * Nagpur Pulse — Decision Review Engine API Client
 * Connects frontend Police Command Center to /api/decision-review
 */

import { apiClient } from './client';

export type AssuranceStatus = 'ASSURED' | 'REVIEW REQUIRED' | 'LOW ASSURANCE' | 'BLOCKED';

export interface DASComponents {
  incident_severity: number;
  traffic_risk: number;
  crime_risk: number;
  event_risk: number;
  unit_availability: number;
  unit_capability: number;
  eta_score: number;
  coverage_safety: number;
  resource_workload: number;
  data_reliability: number;
  prediction_stability: number;
  ml_confidence: number;
}

export interface HardConstraints {
  passed: boolean;
  unit_available: boolean;
  capability_matched: boolean;
  coverage_safe: boolean;
  event_compliant: boolean;
  data_valid: boolean;
  violations: string[];
}

export interface AlternativeUnit {
  unit_id: string;
  callsign: string;
  unit_type: string;
  eta_minutes: number;
  distance_km: number;
  capability_match_pct: number;
  coverage_impact_pct: number;
  what_if_penalty: number;
  decision_assurance_score: number;
  assurance_status: AssuranceStatus;
  preference_reason: string;
}

export interface ConditionDisclosure {
  category: string;
  label: string;
  detail: string;
  verified: boolean;
  source: string;
}

export interface DecisionEvidenceRecord {
  decision_id: string;
  incident_id?: string;
  recommendation_id?: string;
  location_id?: number;
  location_name?: string;
  model_version: string;
  input_snapshot_id: string;
  api_freshness_seconds: number;
  data_reliability_score: number;
  ml_confidence_score: number;
  ml_risk_score: number;
  ml_risk_tier: string;
  recommended_unit: {
    unit_id?: string;
    callsign?: string;
    eta_minutes: number;
    distance_km: number;
  };
  hard_constraints: HardConstraints;
  das_components: DASComponents;
  raw_das_score: number;
  what_if: {
    penalty: number;
    coverage_impact_pct: number;
    details: {
      coverage_loss_pct: number;
      secondary_risk_increase: number;
      resource_impact_pct: number;
      eta_penalty_val: number;
      calculated_penalty: number;
    };
  };
  final_assurance_score: number;
  assurance_status: AssuranceStatus;
  assurance_narrative: string;
  alternatives: AlternativeUnit[];
  known_conditions: ConditionDisclosure[];
  unknown_conditions: ConditionDisclosure[];
  commander: {
    id?: number;
    username?: string;
    role?: string;
    zone?: string;
    action: string;
    override_reason?: string;
    notes?: string;
    final_dispatched_unit_id?: string;
    decision_timestamp?: string;
  };
  outcome: {
    status?: string;
    failure_classification?: string;
    actual_response_time_minutes?: number;
    post_event_evaluation?: string;
    recorded_at?: string;
  };
  audit_chain: {
    sha256_hash: string;
    previous_hash: string;
  };
  created_at?: string;
  updated_at?: string;
}

export interface EvaluateDecisionRequest {
  recommendation_id?: string;
  incident_id?: string;
  location_id?: number;
  location_name?: string;
  recommended_unit_id?: string;
  incident_severity?: number;
  incident_type?: string;
  required_capabilities?: string[];
  min_sector_coverage_pct?: number;
}

export interface CommanderDecisionPayload {
  action: 'APPROVE' | 'MODIFY' | 'REJECT';
  selected_unit_id?: string;
  reason_code?: string;
  comment?: string;
}

export interface OutcomePayload {
  outcome_status: string;
  failure_classification?: 'DATA_FAILURE' | 'MODEL_FAILURE' | 'RECOMMENDATION_FAILURE' | 'HUMAN_DECISION' | 'EXECUTION_FAILURE' | 'NONE';
  actual_response_time_minutes?: number;
  post_event_evaluation?: string;
}

export interface AuditChainVerification {
  verified: boolean;
  total_records: number;
  chain_intact: boolean;
  corrupted_count: number;
  corrupted_records: any[];
  latest_head_hash: string;
  verification_timestamp: string;
}

// ─── Mock Fallback Generator ──────────────────────────────────────────────────

function createMockEvidenceRecord(payload: EvaluateDecisionRequest): DecisionEvidenceRecord {
  const unitId = payload.recommended_unit_id || 'P17';
  const locName = payload.location_name || 'Sitabuldi Interchange';
  const severity = payload.incident_severity || 82;
  const decId = `DEC-${new Date().getFullYear()}-${Math.floor(1000 + Math.random() * 9000)}`;

  const alternatives: AlternativeUnit[] = [
    {
      unit_id: 'P12',
      callsign: 'P12 - Dhantoli Patrol',
      unit_type: 'PATROL',
      eta_minutes: 4.2,
      distance_km: 1.8,
      capability_match_pct: 94.0,
      coverage_impact_pct: 12.5,
      what_if_penalty: 4.2,
      decision_assurance_score: 88.5,
      assurance_status: 'ASSURED',
      preference_reason: 'Fastest alternative backup with minimal coverage depletion in Zone 2',
    },
    {
      unit_id: 'P05',
      callsign: 'P05 - Sadar Quick Response',
      unit_type: 'QRT',
      eta_minutes: 6.8,
      distance_km: 3.1,
      capability_match_pct: 88.0,
      coverage_impact_pct: 18.0,
      what_if_penalty: 6.5,
      decision_assurance_score: 81.2,
      assurance_status: 'REVIEW REQUIRED',
      preference_reason: 'Heavy tactical capability but higher cross-sector transit delay',
    },
    {
      unit_id: 'P21',
      callsign: 'P21 - Sitabuldi Traffic Unit',
      unit_type: 'TRAFFIC',
      eta_minutes: 8.5,
      distance_km: 4.2,
      capability_match_pct: 75.0,
      coverage_impact_pct: 24.0,
      what_if_penalty: 9.1,
      decision_assurance_score: 72.0,
      assurance_status: 'REVIEW REQUIRED',
      preference_reason: 'Traffic specialized unit, higher depletion risk for Sitabuldi corridor',
    },
  ];

  return {
    decision_id: decId,
    incident_id: payload.incident_id || `INC-${Date.now().toString().slice(-4)}`,
    recommendation_id: payload.recommendation_id || `REC-${Date.now().toString().slice(-4)}`,
    location_id: payload.location_id || 1,
    location_name: locName,
    model_version: 'rf_v3_hybrid_ensemble',
    input_snapshot_id: `snap_${Date.now()}`,
    api_freshness_seconds: 14.2,
    data_reliability_score: 94.8,
    ml_confidence_score: 91.5,
    ml_risk_score: severity,
    ml_risk_tier: severity >= 80 ? 'CRITICAL' : severity >= 60 ? 'HIGH' : 'MEDIUM',
    recommended_unit: {
      unit_id: unitId,
      callsign: `${unitId} - Sitabuldi Primary Response`,
      eta_minutes: 3.1,
      distance_km: 1.2,
    },
    hard_constraints: {
      passed: true,
      unit_available: true,
      capability_matched: true,
      coverage_safe: true,
      event_compliant: true,
      data_valid: true,
      violations: [],
    },
    das_components: {
      incident_severity: severity,
      traffic_risk: 78.5,
      crime_risk: 62.0,
      event_risk: 45.0,
      unit_availability: 95.0,
      unit_capability: 92.0,
      eta_score: 91.0,
      coverage_safety: 85.0,
      resource_workload: 74.0,
      data_reliability: 94.8,
      prediction_stability: 96.0,
      ml_confidence: 91.5,
    },
    raw_das_score: 93.4,
    what_if: {
      penalty: 4.8,
      coverage_impact_pct: 12.0,
      details: {
        coverage_loss_pct: 8.5,
        secondary_risk_increase: 3.2,
        resource_impact_pct: 4.0,
        eta_penalty_val: 2.1,
        calculated_penalty: 4.8,
      },
    },
    final_assurance_score: 88.6,
    assurance_status: 'ASSURED',
    assurance_narrative: 'High assurance deployment: unit is within 3.1 min ETA with zero constraint violations and minimal sector coverage risk.',
    alternatives,
    known_conditions: [
      { category: 'TRAFFIC', label: 'Congestion Level', detail: 'Moderate flow on Wardha Rd corridor (Level 2)', verified: true, source: 'Nagpur Smart City ICCC' },
      { category: 'WEATHER', label: 'Visibility & Rain', detail: 'Clear, 28°C, Dry road surface', verified: true, source: 'IMD Nagpur Live Telemetry' },
      { category: 'RESOURCE', label: 'Unit Status', detail: `${unitId} reported idle and patrolling sector`, verified: true, source: 'VHF Radio CAD link' },
    ],
    unknown_conditions: [
      { category: 'EVENT', label: 'VIP Convoy Route', detail: 'Potential unnotified transit near Variety Square in next 30 min', verified: false, source: 'Field Dispatcher Flag' },
      { category: 'CIVIL', label: 'Construction Detour', detail: 'Metro extension work reported on West High Court Rd', verified: false, source: 'Citizen Report' },
    ],
    commander: {
      action: 'PENDING',
    },
    outcome: {
      status: 'PENDING',
    },
    audit_chain: {
      sha256_hash: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
      previous_hash: '0000000000000000000000000000000000000000000000000000000000000000',
    },
    created_at: new Date().toISOString(),
  };
}

// ─── API Functions ──────────────────────────────────────────────────────────────

/**
 * Evaluate an AI recommendation through the 15-step Decision Review Engine.
 */
export async function evaluateRecommendation(payload: EvaluateDecisionRequest): Promise<DecisionEvidenceRecord> {
  try {
    const response = await apiClient<DecisionEvidenceRecord>('/api/decision-review/evaluate', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
    if (response.data) {
      return response.data;
    }
  } catch {
    // fallback
  }
  return createMockEvidenceRecord(payload);
}

/**
 * Fetch a specific decision evidence record by ID.
 */
export async function getDecisionById(decisionId: string): Promise<DecisionEvidenceRecord> {
  const response = await apiClient<DecisionEvidenceRecord>(`/api/decision-review/${decisionId}`);
  if (response.data) {
    return response.data;
  }
  return createMockEvidenceRecord({});
}

/**
 * Fetch recent decision evidence records.
 */
export async function getRecentDecisions(limit = 20, statusFilter?: string): Promise<DecisionEvidenceRecord[]> {
  const query = statusFilter ? `?limit=${limit}&status_filter=${statusFilter}` : `?limit=${limit}`;
  const response = await apiClient<DecisionEvidenceRecord[]>(`/api/decision-review/recent${query}`);
  if (response.data && Array.isArray(response.data)) {
    return response.data;
  }
  return [createMockEvidenceRecord({})];
}

/**
 * Submit commander decision (APPROVE / MODIFY / REJECT).
 */
export async function submitCommanderReviewDecision(
  decisionId: string,
  payload: CommanderDecisionPayload,
): Promise<DecisionEvidenceRecord> {
  try {
    const response = await apiClient<DecisionEvidenceRecord>(`/api/decision-review/${decisionId}/decision`, {
      method: 'POST',
      body: JSON.stringify(payload),
    });
    if (response.data) {
      return response.data;
    }
  } catch {
    // fallback
  }
  const mock = createMockEvidenceRecord({});
  mock.decision_id = decisionId;
  mock.commander = {
    action: payload.action,
    override_reason: payload.reason_code,
    notes: payload.comment,
    final_dispatched_unit_id: payload.selected_unit_id,
    decision_timestamp: new Date().toISOString(),
  };
  return mock;
}

/**
 * Record actual incident outcome and failure taxonomy.
 */
export async function recordDecisionOutcome(
  decisionId: string,
  payload: OutcomePayload,
): Promise<DecisionEvidenceRecord> {
  const response = await apiClient<DecisionEvidenceRecord>(`/api/decision-review/${decisionId}/outcome`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
  if (response.data) {
    return response.data;
  }
  const mock = createMockEvidenceRecord({});
  mock.decision_id = decisionId;
  mock.outcome = {
    status: payload.outcome_status,
    failure_classification: payload.failure_classification,
    actual_response_time_minutes: payload.actual_response_time_minutes,
    post_event_evaluation: payload.post_event_evaluation,
    recorded_at: new Date().toISOString(),
  };
  return mock;
}

/**
 * Verify cryptographic integrity of the SHA-256 tamper-evident hash chain.
 */
export async function verifyAuditHashChain(): Promise<AuditChainVerification> {
  const response = await apiClient<AuditChainVerification>('/api/decision-review/audit-chain/verify');
  if (response.data) {
    return response.data;
  }
  return {
    verified: true,
    total_records: 1,
    chain_intact: true,
    corrupted_count: 0,
    corrupted_records: [],
    latest_head_hash: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
    verification_timestamp: new Date().toISOString(),
  };
}

File: backend/app/schemas/decision_review.py
from typing import List, Dict, Any, Optional
from datetime import datetime
from pydantic import BaseModel, Field


class DASComponentsSchema(BaseModel):
    incident_severity: float = Field(..., ge=0, le=100, description="I - Incident Severity (0-100)")
    traffic_risk: float = Field(..., ge=0, le=100, description="T - Traffic Risk (0-100)")
    crime_risk: float = Field(..., ge=0, le=100, description="C - Crime/Public-Order Risk (0-100)")
    event_risk: float = Field(..., ge=0, le=100, description="E - Event/Crowd Risk (0-100)")
    unit_availability: float = Field(..., ge=0, le=100, description="A - Unit Availability Score (0-100)")
    unit_capability: float = Field(..., ge=0, le=100, description="K - Unit Capability Match (0-100)")
    eta_score: float = Field(..., ge=0, le=100, description="ETA - ETA Rapid Response Score (0-100)")
    coverage_safety: float = Field(..., ge=0, le=100, description="CV - Sector Coverage Safety (0-100)")
    resource_workload: float = Field(..., ge=0, le=100, description="W - Resource Workload Balance (0-100)")
    data_reliability: float = Field(..., ge=0, le=100, description="D - Live API Data Freshness & Quality (0-100)")
    prediction_stability: float = Field(..., ge=0, le=100, description="S - ML Prediction Stability (0-100)")
    ml_confidence: float = Field(..., ge=0, le=100, description="U - ML Posterior Confidence (0-100)")


class HardConstraintCheckSchema(BaseModel):
    passed: bool
    unit_available: bool
    capability_matched: bool
    coverage_safe: bool
    event_compliant: bool
    data_valid: bool
    violations: List[str] = []


class AlternativeUnitSchema(BaseModel):
    unit_id: str
    callsign: str
    unit_type: str
    eta_minutes: float
    distance_km: float
    capability_match_pct: float
    coverage_impact_pct: float
    what_if_penalty: float
    decision_assurance_score: float
    assurance_status: str
    preference_reason: str


class ConditionDisclosureSchema(BaseModel):
    category: str
    label: str
    detail: str
    verified: bool
    source: str


class DecisionReviewEvaluationRequest(BaseModel):
    recommendation_id: Optional[str] = Field(None, description="Recommendation ID if evaluating existing rec")
    incident_id: Optional[str] = Field(None, description="Incident ID if triggered by incident")
    location_id: Optional[int] = Field(None, description="Target junction ID")
    location_name: Optional[str] = Field(None, description="Target junction name")
    recommended_unit_id: Optional[str] = Field(None, description="AI Recommended unit ID")
    incident_severity: Optional[float] = Field(None, ge=0, le=100)
    incident_type: Optional[str] = Field("Road Accident", description="Incident type description")
    required_capabilities: Optional[List[str]] = Field(None, description="Required capabilities e.g. ['AMBULANCE', 'TRAFFIC_PCR']")
    min_sector_coverage_pct: Optional[float] = Field(60.0, description="Minimum allowed sector coverage percentage")


class DecisionReviewEvaluationResponse(BaseModel):
    decision_id: str
    recommendation_id: Optional[str]
    incident_id: Optional[str]
    location_id: Optional[int]
    location_name: Optional[str]
    model_version: str
    api_freshness_seconds: float
    data_reliability_score: float
    ml_confidence_score: float
    ml_risk_score: float
    ml_risk_tier: str
    recommended_unit: Dict[str, Any]
    hard_constraints: HardConstraintCheckSchema
    das_components: DASComponentsSchema
    raw_das_score: float
    what_if_penalty: float
    coverage_impact_pct: float
    final_assurance_score: float
    assurance_status: str
    assurance_narrative: str
    alternatives: List[AlternativeUnitSchema]
    known_conditions: List[ConditionDisclosureSchema]
    unknown_conditions: List[ConditionDisclosureSchema]
    commander_action_required: bool = True
    audit_chain: Dict[str, str]


class CommanderDecisionSubmissionSchema(BaseModel):
    action: str = Field(..., description="APPROVE, MODIFY, or REJECT")
    selected_unit_id: Optional[str] = Field(None, description="Selected police unit ID (required if MODIFY)")
    reason_code: Optional[str] = Field(None, description="Reason code (required if MODIFY or REJECT)")
    comment: Optional[str] = Field(None, description="Commander operational notes")


class OutcomeRecordingSchema(BaseModel):
    outcome_status: str = Field(..., description="SUCCESS, DELAYED, ESCALATED, RESOLVED, FAILED")
    failure_classification: Optional[str] = Field(
        "NONE",
        description="DATA_FAILURE, MODEL_FAILURE, RECOMMENDATION_FAILURE, HUMAN_DECISION, EXECUTION_FAILURE, NONE"
    )
    actual_response_time_minutes: Optional[float] = Field(None, ge=0)
    post_event_evaluation: Optional[str] = Field(None, description="Detailed post-event post-mortem notes")


class DecisionReviewConfigSchema(BaseModel):
    weights: Dict[str, float] = Field(
        default={
            "incident_severity": 0.15,
            "traffic_risk": 0.10,
            "crime_risk": 0.08,
            "event_risk": 0.07,
            "unit_availability": 0.10,
            "unit_capability": 0.08,
            "eta_score": 0.10,
            "coverage_safety": 0.12,
            "resource_workload": 0.05,
            "data_reliability": 0.08,
            "prediction_stability": 0.04,
            "ml_confidence": 0.03,
        }
    )
    what_if_weights: Dict[str, float] = Field(
        default={
            "coverage_loss": 0.40,
            "secondary_risk": 0.30,
            "resource_impact": 0.20,
            "eta_penalty": 0.10,
        }
    )
    thresholds: Dict[str, float] = Field(
        default={
            "assured_min": 85.0,
            "review_required_min": 70.0,
            "low_assurance_min": 50.0,
            "min_sector_coverage": 60.0,
            "max_api_staleness_seconds": 300.0,
        }
    )
